File: exams/gualotuna/u3/exam3/controller/QuickSort.js
export class QuickSort {
    sort(arr) {
      if (arr.length <= 1) return arr;
      const pivot = arr[arr.length - 1];
      const left = [];
      const right = [];
      for (let i = 0; i < arr.length - 1; i++) {
        if (arr[i] < pivot) left.push(arr[i]);
        else right.push(arr[i]);
      }
      return [...this.sort(left), pivot, ...this.sort(right)];
    }
  }
  

File: exams/travez/u3/Exam 3/espe_edu_ec_sorter_controller/Quick_Sort.py
from abc import ABC, abstractmethod
from typing import List

class QuickSort(SortingStrategy):
    
    def sort(self, data: List[int]) -> List[int]:
        arr = data.copy()
        
        def _quick_sort(arr: List[int], low: int, high: int) -> List[int]:
            if low < high:
                pi = _partition(arr, low, high)
                _quick_sort(arr, low, pi-1)
                _quick_sort(arr, pi+1, high)
            return arr
        
        def _partition(arr: List[int], low: int, high: int) -> int:
            pivot = arr[high]
            i = low - 1
            for j in range(low, high):
                if arr[j] < pivot:
                    i += 1
                    arr[i], arr[j] = arr[j], arr[i]
            arr[i+1], arr[high] = arr[high], arr[i+1]
            return i + 1
        
        return _quick_sort(arr, 0, len(arr) - 1)